setup.py

```diff
@@ -66,6 +69,55 @@ def run_tests(self):
         sys.exit(errno)
 
 
+class CMake(BuildExtCommand):
+    def build_extension(self, ext):
+        import shutil
+        import subprocess
+        import tempfile
+
+        to_build = set()
+        # Detect if any source file sits next to a CMakeLists.txt file
+        for source in ext.sources:
+            source_dir = os.path.dirname(os.path.realpath(source))
+            if os.path.exists(os.path.join(source_dir, "CMakeLists.txt")):
+                to_build.add(source_dir)
+
+        if not to_build:
+            # Build the extension as usual
+            BuildExtCommand.build_extension(self, ext)
+            return
+
+        try:
+            cmake_command = os.environ.get("CMAKE_COMMAND", "cmake")
+            build_type = "RelWithDebInfo" if DEBUG_COMPILE else "Release"
+            opts = ["-DCMAKE_BUILD_TYPE={}".format(build_type)]
+            if platform.system() == "Windows":
+                opts.extend(["-A", "x64" if platform.architecture()[0] == "64bit" else "Win32"])
+            else:
+                opts.extend(["-G", "Ninja"])
+                ninja_command = os.environ.get("NINJA_COMMAND", "")
+                if ninja_command:
+                    opts.append("-DCMAKE_MAKE_PROGRAM={}".format(ninja_command))
+
+            for source_dir in to_build:
+                try:
+                    build_dir = tempfile.mkdtemp()
+                    subprocess.check_call([cmake_command, "-S", source_dir, "-B", build_dir] + opts)
+                    subprocess.check_call([cmake_command, "--build", build_dir, "--config", build_type])
+                finally:
+                    if not DEBUG_COMPILE:
+                        shutil.rmtree(build_dir, ignore_errors=True)
+
+            BuildExtCommand.build_extension(self, ext)
+        except Exception as e:
+            if "DD_TESTING_RAISE" in os.environ:
+                raise
+            print('WARNING: building extension "%s" failed: %s' % (ext.name, e))
```

Comment:
hm IIUC this class is going to be used for all extensions, right?

Anyhow, the policy has been to always fail when failing to build an extension so we don't ship/allow to use half-compiled ddtrace.

Reply:
Actually, it only applies to python modules containing a `CMakeLists.txt` file. Other extensions are built (and fail) as usual:
```python
        if not to_build:
            # Build the extension as usual
            BuildExtCommand.build_extension(self, ext)
            return
```

Reply:
Right, though I think I'd stand behind my first command: we want to fail if the build fails, in all cases.

Reply:
👍🏻 to failing if building fails. Since we are able to build wheels for basically everything these days, we control the build step for everyone so we can be sure this doesn't fail.